internal/exporter/exporter.go

```diff
@@ -0,0 +1,178 @@
+	NamespaceIDs []string `json:"namespaceIds"`
+}
+
+type datastoreReqBody struct {
+	OutputURLPrefix string       `json:"outputUrlPrefix,omitempty"`
+	EntityFilter    entityFilter `json:"entityFilter,omitempty"`
+}
+
+type opError struct {
+	Code    int    `json:"code,omitempty"`
+	Message string `json:"message,omitempty"`
+	Status  string `json:"status,omitempty"`
+}
+
+func (e opError) String() string {
+	return fmt.Sprintf("code: %d: %s. Reason: %s", e.Code, e.Status, e.Message)
+}
+
+type operationResp struct {
+	Name     string `json:"name,omitempty"`
+	Metadata struct {
+		OutputURLPrefix string `json:"outputUrlPrefix,omitempty"`
+		Common          struct {
+			StartTime time.Time `json:"startTime,omitempty"`
+			EndTime   time.Time `json:"endTime,omitempty"`
+			State     string    `json:"state,omitempty"`
+		} `json:"common,omitempty"`
+	} `json:"metadata,omitempty"`
+	Done  bool     `json:"done,omitempty"`
+	Error *opError `json:"error,omitempty"`
+}
+
+type Exporter struct {
+	client     *http.Client
+	log        zerolog.Logger
+	ProjectID  string
+	BucketName string
+}
+
+func NewExporter(ctx context.Context, log zerolog.Logger, projectID, bucketName string) (*Exporter, error) {
+	c, err := google.DefaultClient(ctx, scopes...)
+	if err != nil {
+		return nil, fmt.Errorf("error getting default client: %v", err)
+	}
+	return &Exporter{
+		client:     c,
+		log:        log,
+		ProjectID:  projectID,
+		BucketName: bucketName,
+	}, nil
+}
+
+// Export Initiates the datastore export operation exporting all kinds and namespaces
+// Checks to see if the export operation was successful
+func (e *Exporter) Export(ctx context.Context) error {
+	apiUrl := fmt.Sprintf("%sprojects/%s:export", datastoreApi, e.ProjectID)
+
+	b, err := json.Marshal(datastoreReqBody{
+		OutputURLPrefix: "gs://" + e.BucketName,
+		EntityFilter: entityFilter{
+			Kinds:        []string{},
+			NamespaceIDs: []string{},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("error marshalling body: %v", err)
+	}
+	body := bytes.NewReader(b)
+	res, err := e.client.Post(apiUrl, "application/json", body)
+	if err != nil {
+		return fmt.Errorf("error making request: %v", err)
+	}
+	defer res.Body.Close()
+	b, err = ioutil.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+	var r operationResp
+	if err := json.Unmarshal(b, &r); err != nil {
+		return err
+	}
+	if res.StatusCode != 200 {
+		if r.Error != nil {
+			return fmt.Errorf("error making export call: %v", r.Error)
+		}
+		return fmt.Errorf("error making export call: %d: %s", res.StatusCode, res.Status)
+	}
+	e.log.Info().Msgf("Datastore export started. Output location will be: %s", r.Metadata.OutputURLPrefix)
+	return e.watchOp(ctx, r.Name)
+}
+
+func (e *Exporter) interval(dur time.Duration, i chan bool) {
+	for {
+		time.Sleep(dur)
+		i <- true
+	}
+}
+
+func (e *Exporter) watchOp(ctx context.Context, name string) error {
+	interval := make(chan bool)
+	go e.interval(10*time.Second, interval)
```

Comment:
The time standard library provides an implementation of an interval timer that could be used here:
https://gobyexample.com/tickers
In particular this would provide the ability to stop the interval timer, for instance on ctx.Done().

Reply:
ahh thanks for the hint, forgot that existed